### ISSUE TITLE
`license` field in `pyproject.toml` is ignored

### ISSUE THREAD

**Issue by rgommers** [FFY00/meson-python#88]
I'm just noticing this now in the SciPy wheels: the license file specified in `pyproject.toml` is ignored by `meson-python`.

The relevant metadata is:
```toml
[project]
name = "SciPy"
license = {file = "LICENSE.txt"}
```
which should be correct I think, see https://peps.python.org/pep-0621/#license. From that text: _The `file` key has a string value that is a relative file path to the file which contains the license for the project_

Now what is not specified is what should be done with that. What PEP 621 further states is _"The `text` key has a string value which is the license of the project whose meaning is that of the License field from the [core metadata](https://packaging.python.org/specifications/core-metadata/). These keys are mutually exclusive, so a tool MUST raise an error if the metadata specifies both keys."_ Which I think is insufficient, because we need both the license of a package, and the license of individual wheels, which are definitely not the same (wheels can vendor platform-specific components which are under different licenses for example).

In addition to the `license` field, we have the PyPI classifier:
```toml
classifiers = [
    "License :: OSI Approved :: BSD License",
```

Now how should all this work in practice? Here's what I think:

1. The PyPI classifier needs the license of the project itself (fine as is for SciPy).
2. The `Meta > License` field in the left sidebar on the PyPI page (see, e.g., https://test.pypi.org/project/scipy/) needs that same license. But how is it populated, from the classifier or from a `text` field in the PEP 621 license key?
3. The file specified with `license = {file = "LICENSE.txt"}` needs to be included in both sdist and wheels. It does not need to be installed onto the end user's system when the wheel is installed (which makes sense, because there is nowhere for that file to go, it's at the top level in the tree, next to `scipy/` rather than under it).

The above means PEP 621 is wrong. There is an alternative that meets 1-3 above:
- Use `license = {text = "BSD 3-Clause"}` in the metadata, don't specify `file` there.
- Explicitly include the license file by something like `py3.install_sources(['LICENSE.txt'])` in the top-level `meson.build` file
  - The problem with this is that we _don't_ want that file to be installed, but only included in the sdist and wheels. Currently I don't think there is a way to do this?

**Comment by rgommers** [comment]
Note that for other license files in the source tree, we do have to use (and are using) `py3.install_sources(['LICENSE.txt'])` to include them. That's not much of an issue in practice, because they can be installed into `site-packages/scipy` just fine. The issue is mainly with the top-level file, because its install location would be `site-packages/LICENSE.txt`, which is clearly going to give problems.

**Comment by FFY00** [comment]
Yup, ideally, the license text would be included in the metadata itself. This is a known issue, tracked in https://github.com/FFY00/python-pyproject-metadata/issues/3.

Is that solution suitable for you? Or do you really want us to include the license file?

> Use `license = {text = "BSD 3-Clause"}` in the metadata, don't specify `file` there.

The `text` keyword is used to specify the license text, not name.

**Comment by rgommers** [comment]
> Yup, ideally, the license text would be included in the metadata itself.

Do you mean inside the `METADATA` file in the `.dist-info` directory? Right now that contains:
```
Metadata-Version: 2.1
Name: scipy
...
License: BSD
Download-URL: https://github.com/scipy/scipy/releases
```
If so, replacing that with pages and pages of license text does not seem healthy.

At the moment we have a separate `LICENSE.txt` file inside the `.dist-info` directory, next to `WHEEL`, `METADATA`, `RECORD` - that seems a lot nicer. There is also a specific field for this inside `METADATA`, namely:
```
License-File: LICENSE.txt
```

**Comment by rgommers** [comment]
I had a look at this a little more. Different orgs have different guidance, but it's typically something like _"It does not say specifically that the copy of the license you received along with the program is in the file COPYING or LICENSE, but that's where it's usually put."_ (quote from [this guide](https://producingoss.com/en/license-quickstart.html#license-quickstart-applying)).

The PEP 621 idea of putting it as a string in `METADATA` is highly unusual and imho deficient, because no one will think of looking there. I don't see a reason not to stay with the much more standard separate file.

It's also easy to do; I don't understand why https://github.com/FFY00/python-pyproject-metadata/issues/3 is blocked but it should not be relevant. The solution of copying the file given by `license = {file = ...}` into the `.dist-info` can be implemented in a few lines of code in `mesonpy/__init__.py` right next to:
```
            whl.writestr(f'{self.distinfo_dir}/METADATA', self._project.metadata)
            whl.writestr(f'{self.distinfo_dir}/WHEEL', self.wheel)
```
and is a significant improvement over silently dropping the license.

**Comment by FFY00** [comment]
`License-File` is part of [PEP 639](https://peps.python.org/pep-0639/) which is still in draft status, therefore I don't think the copying the license file over to the metadata directory solves the issue, but I do think that's a good idea nevertheless as it somehow mitigates it.

> It's also easy to do; I don't understand why [FFY00/python-pyproject-metadata#3](https://github.com/FFY00/python-pyproject-metadata/issues/3) is blocked but it should not be relevant.

There's a custom format for multiline fields in the metadata file, which is not supported right now. License text are almost always multiline, so to be able to put the license text in the `License` key, we need to fix that first.

**Comment by FFY00** [comment]
This should have been fixed in pyproject-metadata 0.6.0.
